contracts/staking/StakingPool.sol

```diff
@@ -0,0 +1,470 @@
+
+        for (uint256 i = 0; i < rewards.length; i++) {
+            //slither-disable-next-line calls-loop
+            _transferRewards(user.rewardAmounts[i], IERC20(rewards[i].token));
+        }
+    }
+
+    // withdraw when the pool is not going ahead (earlyWithdraw)
+    function withdrawWithoutRewards() external stakingPoolRequirementsUnmet {
+        _withdrawWithoutRewards();
+    }
+
+    function initialize(StakingPoolLib.Data calldata info)
+        external
+        virtual
+        initializer
+    {
+        __RoleAccessControl_init();
+        __Context_init_unchained();
+        __Pausable_init();
+
+        //slither-disable-next-line timestamp
+        require(
+            info.epochStartTimestamp >= block.timestamp,
+            "StakingPool: start time"
+        );
+
+        // todo should we check if stakeToken != any of the reward tokens?
+        require(
+            address(info.stakeToken) != address(0),
+            "StakingPool: stake token defined"
+        );
+        //slither-disable-next-line timestamp
+        require(
+            info.rewardsAvailableTimestamp >
+                info.epochStartTimestamp + info.epochDuration,
+            "StakingPool: init rewards"
+        );
+        require(info.treasury != address(0), "StakePool: nonzero treasury"); // TODO: are we checking if the treasury is whitelisted to that daoId
+        require(!info.emergencyMode, "StakePool: init emergency mode");
+
+        require(info.maxTotalPoolStake > 0, "StakePool: maxTotalPoolStake > 0"); // is this check pointless?
+        require(info.epochDuration > 0, "StakePool: epochDuration > 0"); // is this check pointless?
+        require(info.minimumContribution > 0, "StakePool: minimumContribution"); // is this check pointless?
+        require(info.totalStakedAmount == 0, "StakePool: totalStakedAmount"); // or should we move this to be a single field?
+
+        if (info.launchPaused) {
+            _pause();
+        }
+
+        _stakingPoolInfo = info;
+    }
+
+    function emergencyWithdraw() external emergencyModeEnabled {
+        _withdrawWithoutRewards();
+    }
+
+    function initializeRewardTokens(
+        address treasury,
+        StakingPoolLib.RewardToken[] calldata rewards
+    ) external atLeastDaoMeepleRole(_stakingPoolInfo.daoId) {
+        _initializeRewardTokens(treasury, rewards);
+    }
+
+    function adminEmergencyRewardSweep()
+        external
+        atLeastDaoAdminRole(_stakingPoolInfo.daoId)
+        emergencyModeEnabled
+    {
+        _adminEmergencyRewardSweep();
+    }
+
+    function enableEmergencyMode()
+        external
+        atLeastDaoAdminRole(_stakingPoolInfo.daoId)
+    {
+        _stakingPoolInfo.emergencyMode = true;
+    }
+
+    function setRewardsAvailableTimestamp(uint32 timestamp)
+        external
+        atLeastDaoAdminRole(_stakingPoolInfo.daoId)
+    {
+        _setRewardsAvailableTimestamp(timestamp);
+    }
+
+    function stakingPoolData()
```

Comment:
getter for @CjHare

Reply:
Actually, @octavioamu needs this one for the UI